File: thesis/src/sources/echo/Context01.java
import java.io.*;
import java.util.*;
import java.util.function.*;
public class Context01 implements Closeable {
    private final PrintWriter valueWriter;
    private final PrintWriter exceptionWriter;
    public Context01() throws Exception {
        this.valueWriter = new PrintWriter("raIuhcbB2_values.txt");
        this.exceptionWriter = new PrintWriter("raIuhcbB2_exceptions.txt");
    }
    private void writeSeparator() throws Exception {
        valueWriter.write("--raIuhcbB2-- SEP");
        exceptionWriter.write("--raIuhcbB2-- SEP");
        System.err.print("--raIuhcbB2-- SEP");
        System.out.print("--raIuhcbB2-- SEP");
        valueWriter.flush();
        exceptionWriter.flush();
        System.err.flush();
        System.out.flush();
    }
    private void sendValue(Object value) throws Exception {
        Values.send(valueWriter, value);
    }
    private void sendException(Throwable exception) throws Exception {
        Values.sendException(exceptionWriter, exception);
    }
    private void sendSpecificValue(EvaluationResult value) {
        Values.sendEvaluated(valueWriter, value);
    }
    private void sendSpecificException(EvaluationResult exception) {
        Values.sendEvaluated(exceptionWriter, exception);
    }
    void execute() throws Exception {
        this.writeSeparator();
        try {
            Submission.main(new String[]{});
            sendException(null);
        } catch (Exception | AssertionError e) {
            sendException(e);
        }
    }
    @Override
    public void close() throws IOException {
        this.valueWriter.close();
        this.exceptionWriter.close();
    }
    public static void main(String[] a) throws Exception {
        try(Context01 context = new Context01()) {
            context.execute();
        }
    }
}

File: thesis/src/sources/echo/Context01.hs
{-# LANGUAGE NamedFieldPuns #-}
module Context01 where
import System.IO (hPutStr, stderr, stdout, hFlush)
import System.Environment
import qualified Values
import Control.Monad.Trans.Class
import Control.Exception
import EvaluationUtils
import qualified Submission
value_file = "r23Ae0JOb_values.txt"
exception_file = "r23Ae0JOb_exceptions.txt"
writeSeparator :: IO ()
writeSeparator = do
    hPutStr stderr "--r23Ae0JOb-- SEP"
    hPutStr stdout "--r23Ae0JOb-- SEP"
    appendFile value_file "--r23Ae0JOb-- SEP"
    appendFile exception_file "--r23Ae0JOb-- SEP"
    hFlush stdout
    hFlush stderr
sendValue :: Values.Typeable a => a -> IO ()
sendValue = Values.sendValue value_file
sendException :: Exception e => Maybe e -> IO ()
sendException = Values.sendException exception_file
sendSpecificValue :: EvaluationResult -> IO ()
sendSpecificValue = Values.sendEvaluated value_file
sendSpecificException :: EvaluationResult -> IO ()
sendSpecificException = Values.sendEvaluated exception_file
handleException :: Exception e => (Either e a) -> Maybe e
handleException (Left e) = Just e
handleException (Right _) = Nothing
main = do
    writeSeparator
    let mainArgs = {}
    result <- try (withArgs mainArgs Submission.main) :: IO (Either SomeException ())
    let ee = handleException result in sendException (ee)
    putStr ""